test/e2e/agent/agent_updates_test.go

```diff
@@ -76,6 +76,104 @@ var _ = Describe("VM Agent behavior during updates", func() {
 			// Expect(device.Status.Updated.Status).ToNot(Equal(v1alpha1.DeviceUpdatedStatusType("Unknown")))
 			logrus.Info("Device updated to new image 🎉")
 		})
+
+		It("Should update to v4 with embedded application", Label("updates", "rh-77667"), func() {
+			// Check the device status right after bootstrap
+			response := harness.GetDeviceWithStatusSystem(deviceId)
+			device := response.JSON200
+			Expect(device.Status.Summary.Status).To(Equal(v1alpha1.DeviceSummaryStatusType("Online")))
+			Expect(*device.Status.Summary.Info).To(Equal("Bootstrap complete"))
+			Expect(device.Status.Updated.Status).To(Equal(v1alpha1.DeviceUpdatedStatusType("Unknown")))
+			var newImageReference string
+
+			harness.UpdateDeviceWithRetries(deviceId, func(device *v1alpha1.Device) {
+				currentImage := device.Status.Os.Image
+				logrus.Infof("current image for %s is %s", deviceId, currentImage)
+				repo, _ := parseImageReference(currentImage)
+				newImageReference = repo + ":v4"
+				device.Spec.Os = &v1alpha1.DeviceOSSpec{Image: newImageReference}
+				logrus.Infof("updating %s to image %s", deviceId, device.Spec.Os.Image)
+			})
```

Comment:
This is common code with L34-48, can we make a function that does all that, i.e.

device := WaitForBootstrapAndUpdateToVersion(harness, deviceId, "v4")

Then it will be re-usable for any new additional tests.

Reply:
Well, L49-77 is also common, should also be part of such function.

Reply:
Yes, will done asap